futures-async-macro/src/lib.rs

```diff
@@ -248,6 +248,99 @@ pub fn async(attribute: TokenStream, function: TokenStream) -> TokenStream {
 }
 
 #[proc_macro]
+pub fn async_closure(input: TokenStream) -> TokenStream {
+    let input = TokenStream::from(TokenTree {
+        kind: TokenNode::Group(Delimiter::Parenthesis, input),
+        span: Default::default(),
+    });
+    let paren_group: Expr = syn::parse(input)
+        .expect("failed to parse tokens as an expression");
+
+    let ExprParen {
+        expr,
+        ..
+    } = match paren_group.node {
+        ExprKind::Paren(expr) => expr,
+        _ => panic!("expected contents to be enclosed in parens")
+    };
+
+    let expr = *expr;
+    let Expr { node, .. } = expr;
+
+    let ExprClosure {
+        capture,
+        decl,
+        body,
+        ..
+    } = match node {
+        ExprKind::Closure(expr) => expr,
+        _ => panic!("async_closure! can only be applied to closures"),
+    };
+
+    let FnDecl {
+        inputs,
+        variadic,
+        ..
+    } = { *decl };
+    assert!(!variadic, "variadic functions cannot be async");
+
```

Comment:
this next bit of code is a hack I put in to only accept closures with block bodies at the moment